synapse/handlers/federation_event.py

```diff
@@ -1726,6 +1726,9 @@ async def prep(event: EventBase) -> None:
                             event,
                             auth_event_id,
                         )
+                        # Drop the event from the auth_map too, else we may incorrectly persist
+                        # events which depend on this dropped event.
+                        auth_map.pop(event.event_id, None)
                         return
                     auth.append(ae)
```

Comment:
Additionally, should `check_state_independent_auth_rules` even be using the full `auth_map`?

Feels like we should be constructing a subset map with just the `event.auth_event_ids()`

---

```python
auth_events_for_event = []
auth_event_map_for_event = {}
```

```suggestion
                        # Drop the event from the auth_map too, else we may incorrectly persist
                        # events which depend on this dropped event.
                        auth_map.pop(event.event_id, None)
                        return
                    auth_events_for_event.append(ae)
                    auth_event_map_for_event[auth_event_id] = ae
```

```python
                    await check_state_independent_auth_rules(
                        self._store, event, batched_auth_events=auth_event_map_for_event
                    )
                    check_state_dependent_auth_rules(event, auth_events_for_event)
```

Reply:
It doesn't matter. The `batched_auth_events` param is just a cache of events, so having more in there doesn't hurt in theory.